Menu overhaul

SiverDX opened the pull request:
Replace enchantment menu / screen mixin with own menu

Add scrolling behaviour + the option to increase the max. amount of enchantments

Still WIP
- Need to register menu type for Fabric
- Consider some other stuff

SiverDX committed:
Adjust some mixins
- replace some redirects
- simplify things

SiverDX committed:
Move Mixin logic into own menu / screen
+ Add option to expand max. enchantment amount
+ Add scroll behaviour
- (WIP)

SiverDX committed:
adjust handling of enchantment amount

SiverDX commented:
How it currently looks like

https://streamable.com/8pbvyz

SiverDX commented:
Currently you can add enchantments up to the amount specified in the config 

Source will utilize all 30 (internal reasonable limit - can be increased but not by config) slots to showcase the available ones 

after that table will fill up the slots (up to 3) - table won't add anything if 3 slots are already filled due to the source 

(3 is basically to balance the reroll functionality - you could increase this number or make it a config option)

if the item has reached max enchantments (config value, 3 by default) both won't add any options anymore 

however you will always be able to upgrade the current enchantments even if your item has more than the allowed number (before install, modded or w/e)

Could add some logic to prevent upgrades if you have more than the max. but not too much of a fan of that

SiverDX committed:
register menu on fabric

SiverDX committed:
changes
- adjust fabric mixin name
- allow value of 0 for max enchantment level bound (to disable override)
- (changed some other usages to 'getMaxLevel' since the mixin to that method will refer to the config if relevant)

SiverDX committed:
Merge branch 'suggestions' into menu_overhaul

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/config/Config.java:
```diff
@@ -62,7 +66,7 @@ public class Config {
     // region Enchantment Values
     public static final Property.Bounded<Integer> BOUNDED_ENCHANTMENT_MAX_LEVEL = GROUP_ENCHANTMENT_GENERAL.register(
-            new Property.Bounded<>("enchantment_max_level", "The maximum level an enchantment can be *naturally* obtained at.", GROUP_ENCHANTMENT_GENERAL, 3, 1, 254, Codec.INT)
+            new Property.Bounded<>("enchantment_max_level", "The maximum level an enchantment can be *naturally* obtained at (set it to 0 to disable this override).", GROUP_ENCHANTMENT_GENERAL, 3, 0, 254, Codec.INT)
```
Status: Rejected
Reason: User hidden functionality

Instead register a new `Property.Binary BINARY_OVERRIDE_MAX_ENCHANTMENT_LEVEL` and use that to clearly communicate that this feature can be disabled.

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/api/enchantment/EnchantmentSource.java:
Status: Accepted

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/Common.java:
Status: Accepted

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/enchantment/OverhauledEnchantmentMenu.java:
```diff
@@ -0,0 +1,660 @@
+    public static final int LAPIS_SLOT = 1;
+    public static final int SOURCE_SLOT = 2;
+
+    private static final int LAST_INVENTORYS_SLOT = 29;
```
Typo in `LAST_INVENTORYS_SLOT` -> `LAST_INVENTORY_SLOT`.

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/enchantment/OverhauledEnchantmentMenu.java:
```diff
@@ -0,0 +1,660 @@
+    private static final int LAST_INVENTORYS_SLOT = 29;
+    private static final int LAST_HOTBAR_SLOT = 38;
+
+    private static final ResourceLocation EMPTY_SLOT_LAPIS_LAZULI = ResourceLocation.withDefaultNamespace("item/empty_slot_lapis_lazuli");
```
Unused in mixin version, therefore unnecessary here as well.

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/enchantment/OverhauledEnchantmentMenu.java:
```diff
@@ -0,0 +1,660 @@
+        });
+    }
+
+    private List<EnchantmentInstance> gatherEnchantments(final ItemStack stack, final ItemStack source, final Level level, final BlockPos position) {
```
Wrong priority order: Enchantments must be read
1. first from `Target` (as upgrades),
2. then from `Source` (as override upgrades),
3. then from `Source` (as transfers),
4. and finally from `Table` (as new applications).

This is to ensure that adding more enchantments requires higher experience levels.

In your implementation enchantments are read
1. first from `Source` (as transfers),
2. then from `Table` (as new applications),
3. then from `Target` (as upgrades),
4. then form `Target` (as showcases),
6. and finally from `Source` (as override upgrades).

This causes already applied enchantments to occupy the last slots of the table, allowing players to keep applying enchantments with lower experience level requirements.

This is non-negotiable.

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/enchantment/OverhauledEnchantmentMenu.java:
```diff
@@ -0,0 +1,660 @@
+            return;
+        }
+
+        for (var entry : EnchantmentLib.getEnchantments(stack).entrySet()) {
```
Missing curse omission logic. If being able to upgrade curses was the intentions, it should be handled with a binary config property which defaults to `false`.

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/enchantment/OverhauledEnchantmentMenu.java:
```diff
@@ -0,0 +1,660 @@
+        return enchantments;
+    }
+
+    private void gatherCurrent(final ItemStack stack, final boolean fullyUpgraded, final List<EnchantmentInstance> enchantments) {
```
You suggest implicit sorting determined by the call order via the `fullyUpgraded` boolean flag.

If you really need sorting on the enchantments it should be explicit using a configurable comparator sorting on a method internal temporary list, eliminating the need to call `gatherCurrent` multiple times.

That makes the code more maintainable and easier to understand.

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/enchantment/OverhauledEnchantmentMenu.java:
I understand that this is still a draft, but I want to save you time and keep you from having to rewrite everything. This project does not have an explicit contribution policy but please stick to the following guidelines:

1. Changes should be local and atomic.
    * If you want to introduce a configurable enchantment amount and implement a scrolling feature to the screen, you should not change things which are unrelated to the topic at hand, like the order in which enchantments appear in the menu.
2. Do not rename or replace methods without a very good reason and motivation.
3. Comment your additions and changes, inferring intentions takes time and is unreliable at best. I know that comments lie and code is the truth, but at least knowing what the goal was helps.

Unfortunately I don't have the time to validate any of the changes you suggest to the already existing methods, so I am making the decision to keep all `@Unique` annotated methods as they are. The changes you need to make for the scrolling behavior should remain in the local scope of each method.

JohnSmith474Auth commented on common/src/main/java/johnsmith/enchantingoverhauled/mixin/enchantment/EnchantmentMixin.java:
```diff
@@ -1,34 +1,32 @@
+            return original;
+        }
+
+        int newLevel = Config.BOUNDED_ENCHANTMENT_MAX_LEVEL.get();
```
Again, this should be an explicit config property.

JohnSmith474Auth reviewed (commented):
I understand that this is still a draft, but I want to save you time and keep you from having to rewrite everything. This project does not have an explicit contribution policy but please stick to the following guidelines:

1. Changes should be local and atomic.
    * If you want to introduce a configurable enchantment amount and implement a scrolling feature to the screen, you should not change things which are unrelated to the topic at hand, like the order in which enchantments appear in the menu.
2. Do not rename or replace methods without a very good reason and motivation.
3. Comment your additions and changes, inferring intentions takes time and is unreliable at best. I know that comments lie and code is the truth, but at least knowing what the goal was helps.

Unfortunately I don't have the time to validate any of the changes you suggest to the already existing methods, so I am making the decision to keep all `@Unique` annotated methods as they are. The changes you need to make for the scrolling behavior should remain in the local scope of each method.

JohnSmith474Auth commented:
Also, promoting the EnchantmentMenuMixin and EnchantmentScreenMixin into their own menu and screen will introduce explicit incompatability with [ImmersiveUi](https://modrinth.com/mod/immersive-ui). The item animations will still work, but the enchant animation on the item will be missing, which it does when viewing your WIP:

> How it currently looks like
> 
> https://streamable.com/8pbvyz

SiverDX commented:
> I understand that this is still a draft, but I want to save you time and keep you from having to rewrite everything. This project does not have an explicit contribution policy but please stick to the following guidelines:
> 
> 1. Changes should be local and atomic.
>    
>    * If you want to introduce a configurable enchantment amount and implement a scrolling feature to the screen, you should not change things which are unrelated to the topic at hand, like the order in which enchantments appear in the menu.
> 2. Do not rename or replace methods without a very good reason and motivation.
> 3. Comment your additions and changes, inferring intentions takes time and is unreliable at best. I know that comments lie and code is the truth, but at least knowing what the goal was helps.
> 
> Unfortunately I don't have the time to validate any of the changes you suggest to the already existing methods, so I am making the decision to keep all `@Unique` annotated methods as they are. The changes you need to make for the scrolling behavior should remain in the local scope of each method.

I understand that there are a lot of changes but the current repository and insane mixins into vanilla code to basically add your custom menus and whatnot without defining your own is unsustainable
- It eliminates any chance of contribution
- Is brittle in terms of future updates from Mojang (changes to the base classes that you won't initially notice / breaking changes)
- Some mixins / changes from other mods may apply to the base classes while other parts are ignored due to inject&cancel (= override), resulting in inconsistent behaviour (and potentially crashes)

This PR (and the other PR) is not just about the scrolling behaviour but also trying to provide some cleaner implementations, especially in terms of trying to clean up the mixin usage to make it more readable / understandable

These points may not be important to you but at that point you'd be better of just closing these PRs

If not I can try to change the names to match more closely with the original counterparts / do other adjustments
Because I heavily advise against mixin the original implementation (`@Uniue` annotated parts) with this overhaul

---

Things like the order were initially adjusted so that relevant entries appear in the top, since now you could have scrollable entries
And for a user the first entries would be the table enchantments since those are the ones that can be rerolled
(It is also why the `gatherCurrent` was split up - you'd want non-interactable entries to be at the bottom)

I'd assume another change would be needed where entries are sorted after gathering or sth. like that
Since you'd only know if an enchantment is interactable after the override has been applied

For now I adjusted the sorting to match the existing order

---

I'd recommend handling compatibility in a more clear format
By adding specific support for them

Regarding ImmersiveUI I'd create  a PR to them that moves the code from the mixin into a Utils method, making it re-usable
And then applying it to the `OverEnchantmentScreen` via self-mixin if said mod is loaded
(Or add it directly into the class, guarded by a mod check)

---

I agree that documentation is important and I plan to add some additional parts
Though I'll have to say that the comments in the current repository mostly result in

```java
// Flags whether person is old
boolean isOld = true;
```

JohnSmith474Auth commented:
Edit: My shift button gave out while trying to enter a new line and I ended up commenting prematurely.

> > I understand that this is still a draft, but I want to save you time and keep you from having to rewrite everything. This project does not have an explicit contribution policy but please stick to the following guidelines:
> > 
> > 1. Changes should be local and atomic.
> >    
> >    * If you want to introduce a configurable enchantment amount and implement a scrolling feature to the screen, you should not change things which are unrelated to the topic at hand, like the order in which enchantments appear in the menu.
> > 2. Do not rename or replace methods without a very good reason and motivation.
> > 3. Comment your additions and changes, inferring intentions takes time and is unreliable at best. I know that comments lie and code is the truth, but at least knowing what the goal was helps.
> > 
> > Unfortunately I don't have the time to validate any of the changes you suggest to the already existing methods, so I am making the decision to keep all `@Unique` annotated methods as they are. The changes you need to make for the scrolling behavior should remain in the local scope of each method.
> 
> I understand that there are a lot of changes but the current repository and insane mixins into vanilla code to basically add your custom menus and whatnot without defining your own is unsustainable
> 
>   * It eliminates any chance of contribution
> 
>   * Is brittle in terms of future updates from Mojang (changes to the base classes that you won't initially notice / breaking changes)
> 
>   * Some mixins / changes from other mods may apply to the base classes while other parts are ignored due to inject&cancel (= override), resulting in inconsistent behaviour (and potentially crashes)
> 
> 
> This PR (and the other PR) is not just about the scrolling behaviour but also trying to provide some cleaner implementations, especially in terms of trying to clean up the mixin usage to make it more readable / understandable
> 
> These points may not be important to you but at that point you'd be better of just closing these PRs
> 
> If not I can try to change the names to match more closely with the original counterparts / do other adjustments Because I heavily advise against mixin the original implementation (`@Uniue` annotated parts) with this overhaul

We seem to have a misunderstanding. I am not opposed to the idea of moving the menu and screen mixins to their own menu and screen classes. What I meant by "keeping the `@Unique` annotated methods" is migrating them as they are to the their own files.

Your draft completely rewrites the menus functionality, changing multiple things unrelated to the scrolling feature. My comment:
> Also, promoting the EnchantmentMenuMixin and EnchantmentScreenMixin into their own menu and screen will introduce explicit incompatability with [ImmersiveUi](https://modrinth.com/mod/immersive-ui). The item animations will still work, but the enchant animation on the item will be missing, which it does when viewing your WIP:
> 
> > How it currently looks like
> > https://streamable.com/8pbvyz

was not meant as an argument against moving them to their own classes, but just an observation.

What I cannot afford to do right now is spend my time trying to understand the major refactors you made and how the current code base relates and maps to your PR, so in that sense, less is more.


> Things like the order were initially adjusted so that relevant entries appear in the top, since now you could have scrollable entries And for a user the first entries would be the table enchantments since those are the ones that can be rerolled (It is also why the `gatherCurrent` was split up - you'd want non-interactable entries to be at the bottom)
> 
> I'd assume another change would be needed where entries are sorted after gathering or sth. like that Since you'd only know if an enchantment is interactable after the override has been applied
> 
> For now I adjusted the sorting to match the existing order

I would like to keep the menu as it is right now. Making the amount of enchantments configurable is acceptable. Changing the order in which enchantments are displayed is not. Let's establish a scope for this PR: "Making the amount of enchantments confiugrable" and try to stick to it.

> I'd recommend handling compatibility in a more clear format By adding specific support for them
> 
> Regarding ImmersiveUI I'd create a PR to them that moves the code from the mixin into a Utils method, making it re-usable And then applying it to the `OverEnchantmentScreen` via self-mixin if said mod is loaded (Or add it directly into the class, guarded by a mod check)

That is certainly a much better approach to compatibility. Again, just an observation, not an oppositional argument.

> I agree that documentation is important and I plan to add some additional parts Though I'll have to say that the comments in the current repository mostly result in
> 
> ```java
> // Flags whether person is old
> boolean isOld = true;
> ```

That is true, but I don't require api level documentation. All I want is to know why a change is seen as necessary and what the goal of the implementation is. The smaller file changes were manageable, but the massive refactors require justification.

SiverDX committed:
misc
- adjust variable names to previous
- fix enchantment cost in creative
- adjust gathering order / sort existing
- fix wrong parameter order for upgrade enchant
- fix spelling mistake in variable

SiverDX commented:
I understood what you meant with the `@Unique` part

I still advise the same
If the menu and screen are moved to their own classes then use the opportunity to clean up the current code in those parts

Changes are going to be needed either way, due to:
- adjustments how many slots can exist
- scrolling
- changes to the EnchantmentSource class
- direct access to the menu (no accessor)

---

Generally the goal would be to keep the functionality the same, even if the code itself looks a bit different
I'm still testing and try to adjust if I notice functional changes

The exception was the sorting part - not for balancing reason but rather user experience
Which is now changed due to the scrolling / available slot adjustment

And for that I still recommend rethinking how you'd want to display the entries and in which order
The cost can always be adjusted so that new enchantments can appear first or last while still retaining the same cost-balance as the existing behaviour

SiverDX committed:
add some comments for menu

SiverDX committed:
add some comments for screen
+ move tooltip logic into separate methods

SiverDX committed:
build had issues with two annotation tags (?)
